File: backend/routers/normalization.py
"""归一化路由 — 将某分类所有首饰的 usage_count 等比例缩减"""

from fastapi import APIRouter, HTTPException

from backend.database import get_cursor
from backend.models.schemas import NormalizationOut

router = APIRouter(prefix="/api/categories", tags=["normalization"])


@router.post("/{category_id}/normalize", response_model=NormalizationOut)
def normalize_category(category_id: int):
    """归一化：将该分类所有首饰 usage_count 减去当前最小值（需全部 ≥ 1）"""
    with get_cursor(commit=True) as cur:
        # 1. 验证分类存在
        cur.execute("SELECT id FROM categories WHERE id = %s", (category_id,))
        if not cur.fetchone():
            raise HTTPException(404, f"分类 {category_id} 不存在")

        # 2. 获取 usage_count 最小值 + 验证归一化条件
        cur.execute(
            "SELECT MIN(usage_count) AS min_usage, COUNT(*) AS cnt "
            "FROM items WHERE category_id = %s",
            (category_id,),
        )
        stats = cur.fetchone()

        if stats["cnt"] == 0:
            raise HTTPException(400, "该分类下没有首饰，无法归一化")
        if stats["min_usage"] < 1:
            raise HTTPException(400, "存在未佩戴过的首饰，无法归一化")

        min_removed = stats["min_usage"]

        # 3. 所有首饰减去最小值
        cur.execute(
            "UPDATE items SET usage_count = usage_count - %s "
            "WHERE category_id = %s",
            (min_removed, category_id),
        )

        # 4. 记录归一化历史
        cur.execute(
            "INSERT INTO normalizations (category_id, min_removed) VALUES (%s, %s)",
            (category_id, min_removed),
        )
        norm_id = cur.lastrowid

        cur.execute(
            "SELECT id, category_id, min_removed, created_at "
            "FROM normalizations WHERE id = %s",
            (norm_id,),
        )
        row = cur.fetchone()

    return NormalizationOut(**row)

File: frontend/src/db/services/normalization.ts
/** 归一化 service — 替代 normalization.py router
 *
 * 逻辑: 将该分类所有首饰 usage_count 减去当前最小值
 * 条件: 分类下 > 0 件首饰，且所有首饰 usage_count >= 1
 */

import type { Normalization } from '../../api/types';
import { getDBSync, markDirty } from '../database';

export function normalizeCategory(categoryId: number): Normalization {
  const db = getDBSync();

  // 1. 验证分类存在
  const catStmt = db.prepare('SELECT id FROM categories WHERE id = :id');
  catStmt.bind({ ':id': categoryId });
  if (!catStmt.step()) {
    catStmt.free();
    throw new Error(`分类 ${categoryId} 不存在`);
  }
  catStmt.free();

  // 2. 获取 min usage_count 和 count
  const statsStmt = db.prepare(
    'SELECT MIN(usage_count) AS min_usage, COUNT(*) AS cnt FROM items WHERE category_id = :id',
  );
  statsStmt.bind({ ':id': categoryId });
  statsStmt.step();
  const stats = statsStmt.getAsObject();
  statsStmt.free();

  if ((stats.cnt as number) === 0) {
    throw new Error('该分类下没有首饰，无法归一化');
  }
  if ((stats.min_usage as number) < 1) {
    throw new Error('存在未佩戴过的首饰，无法归一化');
  }

  const minRemoved = stats.min_usage as number;

  // 3. 所有首饰减去最小值
  db.run('UPDATE items SET usage_count = usage_count - :m WHERE category_id = :id', {
    ':m': minRemoved,
    ':id': categoryId,
  });

  // 4. 记录归一化历史
  db.run('INSERT INTO normalizations (category_id, min_removed) VALUES (:cid, :m)', {
    ':cid': categoryId,
    ':m': minRemoved,
  });
  const normId = db.exec('SELECT last_insert_rowid() AS id')[0].values[0][0] as number;

  markDirty();

  // 5. 查询返回
  const normStmt = db.prepare(
    'SELECT id, category_id, min_removed, created_at FROM normalizations WHERE id = :id',
  );
  normStmt.bind({ ':id': normId });
  normStmt.step();
  const row = normStmt.getAsObject();
  normStmt.free();

  return {
    id: row.id as number,
    category_id: row.category_id as number,
    min_removed: row.min_removed as number,
    created_at: row.created_at as string,
  };
}
